format.js

```diff
@@ -0,0 +1,386 @@
+// SPDX-FileCopyrightText: 2022 Andre 'Staltz' Medeiros
+//
+// SPDX-License-Identifier: LGPL-3.0-only
+
+const bencode = require('bencode')
+const BFE = require('ssb-bfe')
+const ssbKeys = require('ssb-keys')
+const SSBURI = require('ssb-uri2')
+
+const CONTENT_SIG_PREFIX = Buffer.from('bendybutt', 'utf8')
+
+const feedFormat = {
+  name: 'bendybutt-v1',
+  encodings: ['js'],
+
+  getFeedId(nativeMsg) {
+    return BFE.decode(bencode.decode(nativeMsg, 2, 39))
+  },
+
+  _msgIdCache: new Map(),
+  _extractCache: new Map(),
+
+  getMsgId(nativeMsg) {
+    if (feedFormat._msgIdCache.has(nativeMsg)) {
+      return feedFormat._msgIdCache.get(nativeMsg)
+    }
+    let data = ssbKeys.hash(nativeMsg)
+    if (data.endsWith('.sha256')) data = data.slice(0, -'.sha256'.length)
+    const msgId = SSBURI.compose({
+      type: 'message',
+      format: 'bendybutt-v1',
+      data,
+    })
+    feedFormat._msgIdCache.set(nativeMsg, msgId)
+    return msgId
+  },
+
+  getSequence(nativeMsg) {
+    const { sequence } = feedFormat._extract(nativeMsg)
+    return sequence
+  },
+
+  isNativeMsg(x) {
+    if (!Buffer.isBuffer(x)) return false
+    if (x.length === 0) return false
+    try {
+      const { authorBFE } = feedFormat._extract(x)
+      return BFE.isEncodedFeedBendybuttV1(authorBFE)
+    } catch (err) {
+      return false
+    }
+  },
+
+  isAuthor(author) {
+    return typeof author === 'string' && SSBURI.isBendyButtV1FeedSSBURI(author)
+  },
+
+  toPlaintextBuffer(opts) {
+    const { content, contentKeys, keys, hmacKey } = opts
+    const contentBFE = BFE.encode(content)
+    const contentSignature = ssbKeys.sign(
+      contentKeys || keys,
+      hmacKey,
+      Buffer.concat([CONTENT_SIG_PREFIX, bencode.encode(contentBFE)])
+    )
+    const contentSection = [content, contentSignature]
+    return bencode.encode(BFE.encode(contentSection))
+  },
+
+  _extract(nativeMsg) {
+    if (feedFormat._extractCache.has(nativeMsg)) {
+      return feedFormat._extractCache.get(nativeMsg)
+    }
+    const [payload, signatureBFE] = bencode.decode(nativeMsg)
+    const [authorBFE, sequence, previousBFE, timestamp, contentSection] =
+      payload
+    const extracted = {
+      payload,
+      signatureBFE,
+      authorBFE,
+      sequence,
+      previousBFE,
+      timestamp,
+      contentSection,
+    }
+    feedFormat._extractCache.set(nativeMsg, extracted)
+    return extracted
+  },
+
+  newNativeMsg(opts) {
+    const author = opts.keys.id
+    const previous = opts.previous || { key: null, value: { sequence: 0 } }
+    const sequence = previous.value.sequence + 1
+    const previousId = previous.key
+    const timestamp = +opts.timestamp
+    const content = opts.content
+    const contentBFE = BFE.encode(content)
+    const contentSignature = ssbKeys.sign(
+      opts.contentKeys || opts.keys,
```

Comment:
should we maybe error if contentKeys is not supplied? This could be a footgun

Reply:
I vaguely recall we discussed this in ngi-pointer, and that enforcing this is to be done in ssb-meta-feeds, not ssb-bendy-butt. And then we updated the bendy-butt-spec to say "The signature uses some cryptographic keypair which MAY be different from the author keypair". So a "may" not a "must".

On one hand it's easy for me to add a one line of code that checks for contentKeys or throws, but on the other hand this doesn't pass tests in ssb-feed-format `check()` anymore, because `check()` internally passes some `opts` to `newNativeMsg` and those opts are missing `opts.contentKeys`.

Reply:
Oh now i see it was this way before af well. Let's keep to then